Ullaakut/Bloggo
# Add persistence and use real data models

## Effects of this PR

![bloggomysqllight](https://user-images.githubusercontent.com/6976628/43611562-d02f89ce-96a9-11e8-80ff-0bf6542360c4.gif)


* Adds MySQL connector & repositories
* Add docker-compose to run mysql & adminer
* Add error types for better handling of db errors
* Export logger to specific package & inject logger in every component
* Add user data model
* Update authentication to return user ID & store it in context
* Add feature to force set author as token user ID
* Set log level to debug while features are in progress
* Add multiple debug logs thoughout the code

Fixes #10 
Starts #7 by adding a `docker-compose.yml` file for MySQL and Adminer

## Goal of this PR

Bloggo couldn't be considered a blogging API without a persistent data storage to keep its users and blog posts.

This PR allows a user with a valid admin token to create and get blog articles.

## To do & known issues

- [x] Currently, the `created_at` field isn't set properly (will fix in this PR)
- [x] I need to add a check when getting a user ID to make sure that the user has admin rights (will fix in this PR)
- [x] The route to create a blog post should returned the created entity, containing its attributed ID. (will fix in this PR)

* Need to add validation to blog posts (#23)

## How to test

Since the DB doesn't start automatically populated, it's currently kind of annoying to set it up for testing.

1. Run `docker-compose up` from the root of the repository
2. Visit `0.0.0.0:8080` in your favorite browser
3. Create a `blog_posts` table and a `users` table like such described in the screenshots attached below
4. Build and run Bloggo (`go build` / `go run`)
5. Execute the following command

```
curl -X POST \
       http://0.0.0.0:4242/posts \
       -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6ImJyZW5kYW4ubGUtZ2xhdW5lYyt0ZXN0YXBpQGVwaXRlY2guZXUiLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImlzcyI6Imh0dHBzOi8vc2FtcGxlcy5hdXRoMC5jb20vIiwic3ViIjoiYXV0aDB8NTk2ZjI3YzJjMzcwOTY2MWU5Y2VhMzdkIiwiYXVkIjoia2J5dUZEaWRMTG0yODBMSXdWRmlhek9xak8zdHk4S0giLCJleHAiOjE2MDA0OTI5NjUsImlhdCI6MTUwMDQ1Njk2NX0.4To8mYgu4pM7J2G5jBTnKWelCTU1U1jo0QOENVp3pOk' \
       -H 'Cache-Control: no-cache' \
       -H 'Content-Type: application/json' \
       -d '{
              "title": "Lorem ipsum",
              "content": "Dolor sit amet"
       }'
```

6. Check the ID of the blog post you created in the database with adminer, and run this command while replacing `42` with the ID of the post you want to get

```
curl -X GET \
       http://0.0.0.0:4242/posts/20 \
       -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6ImJyZW5kYW4ubGUtZ2xhdW5lYyt0ZXN0YXBpQGVwaXRlY2guZXUiLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImlzcyI6Imh0dHBzOi8vc2FtcGxlcy5hdXRoMC5jb20vIiwic3ViIjoiYXV0aDB8NTk2ZjI3YzJjMzcwOTY2MWU5Y2VhMzdkIiwiYXVkIjoia2J5dUZEaWRMTG0yODBMSXdWRmlhek9xak8zdHk4S0giLCJleHAiOjE2MDA0OTI5NjUsImlhdCI6MTUwMDQ1Njk2NX0.4To8mYgu4pM7J2G5jBTnKWelCTU1U1jo0QOENVp3pOk'
```

* Try to get blog posts that don't exist
* Try to create invalid blog posts with invalid data
* Try to read from a non-admin account (should pass)
* Try to write from a non-admin account (should fail)
* Make sure that when creating an entity, the API returns its data with its attributed ID

<img width="440" alt="screenshot 2018-08-02 at 23 14 48" src="https://user-images.githubusercontent.com/6976628/43611581-e4bf24f8-96a9-11e8-8d03-42d539fcd0aa.png">
<img width="437" alt="screenshot 2018-08-02 at 23 14 56" src="https://user-images.githubusercontent.com/6976628/43611582-e4d921a0-96a9-11e8-90e9-b2bb1b4a0c2e.png">
